### PR TITLE
[Core] Assign input proceess

### LINKED ISSUE #6950: Structural Mechanics Application: reading external loads from text files
Good morning @RiccardoRossi  and @loumalouomega!  I was wondering if there is any elegant way of prescribing external loading from the input files. 
Basically, I have text file with pressure histories that should be applied as input (external forces) at different locations of the structural model (membrane in my case).

Thanks a lot!
Pavel

### PR DESCRIPTION
**Pull request by loumalouomega**
Fixes https://github.com/KratosMultiphysics/Kratos/issues/6950. Sorry for the delay @playpaolo, I needed to do some changes before prepare this PR and find the time to do this

This PR adds a process which allows to assign values from an input (txt or json, see examples in tests). The txt follows the structure ask by @playpaolo, the json is defined by myself

Uses the same class as the FromJSONCheckValues in order to interpolate values on time. 

It is implemented general so can be used both in elements, conditions and nodes (both historical and non-historical values)

It is possible to define the input in an arbitrary geometric point as asked by @playpaolo, but also an Id of a already existing entity can be considered

For now assigns the value from the closest point, but I was thinking in generate an auxiliar mesh to interpolate the values between sources, but this requires further discussion and it is out of the scope (IMO) of the features asked by @playpaolo 

In order to use it, as a regular process:

~~~json

        {
            "process_list" : [
                {
                    "python_module"   : "assign_scalar_input_to_nodes_process",
                    "kratos_module" : "KratosMultiphysics",
                    "process_name"          : "AssignScalarInputToNodesProcess",
                    "Parameters"            : {
                        "model_part_name": "Main",
                        "variable_name"  : "PRESSURE",
                        "historical"     : true,
                        "file"           : "auxiliar_files_for_python_unittest/test_processes/input_geometry.txt"
                    }
                }
                ]
            }
~~~

The structure of the txt file is as follows (separated by \t tabulations):

~~~csv
time 	(0.75, 0.75, 0.0) 	(0.25, 0.25, 0.0)
0.1 	1.0 	2.0
0.2 	2.0 	4.0
0.3 	4.0 	8.0
0.4 	8.0 	16.0
0.5  	16.0 	32.0
~~~

If instead of geometric coordinates, you want to consider ids of entities, just write instead of the coordinates the id:

~~~csv
time 	2 	6
0.1 	1.0 	2.0
0.2 	2.0 	4.0
0.3 	4.0 	8.0
0.4 	8.0 	16.0
0.5  	16.0 	32.0
~~~

For json:

~~~json
{
    "TIME" : [0.1,0.2,0.3,0.4,0.5],
    "1"    : {
        "COORDINATES" : [0.75, 0.75, 0.0],
        "VALUES"      : { "PRESSURE" : [1.0,2.0,4.0,8.0,16.0] }
    },
    "2"    : {
        "COORDINATES" : [0.25, 0.25, 0.0],
        "VALUES"      : { "PRESSURE" : [2.0,4.0,8.0,16.0,32.0] }
    }
}
~~~

for ids:

~~~json
{
    "TIME" : [0.1,0.2,0.3,0.4,0.5],
    "1"    : {
        "ID" : 2,
        "VALUES"      : { "PRESSURE" : [1.0,2.0,4.0,8.0,16.0] }
    },
    "2"    : {
        "ID" : 6,
        "VALUES"      : { "PRESSURE" : [2.0,4.0,8.0,16.0,32.0] }
    }
}
~~~

@KratosMultiphysics/interface-committee FYI

### CODE REVIEW

**Comment by playpaolo** [comment]
Thank you very much, @loumalouomega ! Its exactly what I needed.

**Comment by loumalouomega** [comment]
@playpaolo any suggestion complain?

Any comment from the @KratosMultiphysics/technical-committee ?

**Comment by playpaolo** [comment]
Hola Vicente,
Perdona por no haberte contestado antes. Tengo la semana de los exámenes finales, por lo que estaba un poco "out of" testing. Déjame un par de días más por favor para que pueda darte un feedback. 
Disculpa de nuevo.
Pavel

Enviado desde Yahoo Mail con Android 
 
  El sáb., jun. 20, 2020 a 21:45, Vicente Mataix Ferrándiz<notifications@github.com> escribió:   


@playpaolo any suggestion complain?

Any comment from the @KratosMultiphysics/technical-committee ?

**Comment by loumalouomega** [comment]
> Hola Vicente, Perdona por no haberte contestado antes. Tengo la semana de los exámenes finales, por lo que estaba un poco "out of" testing. Déjame un par de días más por favor para que pueda darte un feedback.  

No problem

**Comment by philbucher** [comment]
Guys english would be appreciated :D

**Comment by loumalouomega** [comment]
@playpaolo does this fit your needs?

File: kratos/processes/assign_scalar_input_to_entities_process.h [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+
+    const Variable<double>* mpVariable = nullptr;                            /// The pointer of the variable
+
+    ResultDatabase mDatabase;                                                /// The database containing the information to assign the values
+
+    std::vector<std::unordered_map<IndexType, double>> mWeightExtrapolation; /// This maps specifies the extrapolation weights for each entity
+
+    std::vector<array_1d<double, 3>> mCoordinates;                           /// The coordinates of the database
+
+    Algorithm mAlgorithm = Algorithm::NEAREST_NEIGHBOUR;                     /// The algorithm considered
```
Comment (philbucher) [inline kratos/processes/assign_scalar_input_to_entities_process.h:201]: Shouldn’t this fit directly? If not then one could use the mappers to fit the results 
I don’t like reimplementing the same functionality

Comment (loumalouomega) [reply kratos/processes/assign_scalar_input_to_entities_process.h:201]: It is not a mapping (unless you create an auxiliar mesh), you give a coordinates and it assigns values according to that into the real mesh

Comment (philbucher) [reply kratos/processes/assign_scalar_input_to_entities_process.h:201]: Hm but why not creating a temp aux mesh?
Then you can use all the available mappers if the mesh is not matching 

This way would even work in mpi (using the mappers from the mapping App)

Comment (loumalouomega) [reply kratos/processes/assign_scalar_input_to_entities_process.h:201]: I was thinking about this in a future PR, this is the first version

Comment (ryzhakov82) [reply kratos/processes/assign_scalar_input_to_entities_process.h:201]: For the tasks we need to perform now it serves perfectly the way it is.

**Comment by ryzhakov82** [comment]
> Guys english would be appreciated :D

Sorry, wasnt aware that its going here (thought it was a private message to Vicente)

**Comment by ryzhakov82** [comment]
> @playpaolo does this fit your needs?

absolutely!, thank you!

**Review by philbucher — APPROVED** [review]
I didn't give this a thorough review but looks ok to me

Approving under the condition that if more interpolation methods become necessary we should think again if mappers can be used to avoid code duplication

**Comment by philbucher** [comment]
@loumalouomega there is a minor issue with using the loop var `i` twice:
https://github.com/KratosMultiphysics/Kratos/blob/512d8f489504e23d4dfc736d8d64734a54b8ba6b/kratos/processes/assign_scalar_input_to_entities_process.cpp#L604-L616

**Pull request merged**